### PR TITLE
Remove old epochs leader selection from memory

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR addresses a TODO from https://github.com/onflow/flow-go/pull/157 to remove old epochs such that we only ever store at most 2 epochs worth of leader selection.

### CODE REVIEW

File: consensus/hotstuff/committees/leader/leader_selection.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func (l LeaderSelection) FirstView() uint64 {
 }
 
 func (l LeaderSelection) FinalView() uint64 {
-	return l.firstView + uint64(len(l.leaderIndexes))
+	return l.firstView + uint64(len(l.leaderIndexes)) - 1
```
Comment (jordanschalm) [inline consensus/hotstuff/committees/leader/leader_selection.go:51]: This fixes an off-by-one error in the original PR, where we incorrectly extended the `FinalView` one view too far into the future. This is easy to see by considering an epoch with only one view.

**Review by AlexHentschel — APPROVED** [review]
Looks great. 

A couple suggestions: 
* I think we should hold 3 epochs:
   - we always need the current Epoch
   - the last Epoch is potentially needed for adjudicating slashing challenges
   - upcoming Epoch 
* This is pure aesthetics: I think you could shorten the entire pruning to 
   ```golang 
	for c := range c.leaders {
		if c + 3 < counter {
		      delete(c.leaders, c)
		}
	}   
   ```
    (apparently, [removing selected keys from map within a range loop is safe in go](https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop))

File: consensus/hotstuff/committees/leader/leader_selection_test.go [hunk 2]
```diff
@@ -96,6 +98,63 @@ func TestDeterministic(t *testing.T) {
...
+	// views before first view should return error
+	t.Run("before first view", func(t *testing.T) {
+		before := rand.Uint64() % firstView
+		_, err = leaders.LeaderForView(before)
+		assert.Error(t, err)
+	})
+
+	// views after final view should return error
+	t.Run("after final view", func(t *testing.T) {
+		after := finalView + uint64(rand.Uint32()) + 1
```
Comment (zhangchiqing) [inline consensus/hotstuff/committees/leader/leader_selection_test.go:152]: better to also test the boundary specifically, which is what the fix was for:
```
               firstViewInNextEpoch := finalView + 1
		_, err = leaders.LeaderForView(after)
		assert.Error(t, err)

                lastViewInPreviousEpoch := firstView - 1
		_, err = leaders.LeaderForView(lastViewInPreviousEpoch)
		assert.Error(t, err)

```

File: consensus/hotstuff/committees/consensus_committee_test.go [hunk 3]
```diff
@@ -114,6 +114,111 @@ func TestConsensus_LeaderForView(t *testing.T) {
...
+		_, has1 := committee.leaders[epochCounter]
+		_, has2 := committee.leaders[epochCounter+1]
+		_, has3 := committee.leaders[epochCounter+2]
+		_, has4 := committee.leaders[epochCounter+3]
+		assert.False(t, has1)
+		assert.True(t, has2)
+		assert.True(t, has3)
+		assert.True(t, has4)
+	})
+}
```
Comment (zhangchiqing) [inline consensus/hotstuff/committees/consensus_committee_test.go:220]: would be great to have a test that use a loop to add 10 epoch transition and in the end assert there are 3 leader selections

**Review by zhangchiqing — APPROVED** [review]
Nice work, some minor suggestions

File: consensus/hotstuff/committees/leader/leader_selection_test.go (hunk 2 above)
Comment (jordanschalm) [reply consensus/hotstuff/committees/leader/leader_selection_test.go]: Done in https://github.com/onflow/flow-go/pull/191/commits/cc9057651d6c0cef9b07d6ab8cb68c9e2e754d87

File: consensus/hotstuff/committees/consensus_committee_test.go (hunk 3 above)
Comment (jordanschalm) [reply consensus/hotstuff/committees/consensus_committee_test.go:179]: 👍 Refactored this test to loop through 10 epochs https://github.com/onflow/flow-go/pull/191/commits/50576f3ae80a231ae2d843a82ea0e62c165b68ac

**Pull request merged**